passes/source2il.nim

```diff
@@ -99,6 +108,22 @@ proc evalType(c; t; n: NodeIndex): SemType =
           discard "all good"
 
       tup
+  of UnionTy:
+    var list = newSeq[SemType]()
+    for i, it in t.pairs(n):
+      # make sure to add the types in an ordered fashion, so that
+      # ``union(int, float)`` and ``union(float, int)`` result in the same type
+      let
+        typ = c.expectNot(c.evalType(t, it), tkVoid)
+        at = lowerBound(list, typ, cmp)
+
+      if at < list.len and list[at] == typ:
+        if typ.kind != tkError:
+          c.error("union type operands must be unique")
```

Comment:
>Perhaps I'm wrong, just looking at this but I think `union(union(float, int), union(int, string))` should result in `union(float, int, string)`.

I did think about this when writing the specification. My main motivation was to allow `int or float or string` producing the `union(int, float, string)` type instead of `union(union(int, float), string)` without any special cases or rules.

I ultimately discarded the "flatten" idea, mostly for practical usage reasons, but also because it means that type formation, term elimination, and term introduction rules become more complex for `union`.

There are two concrete situations where I think flattening is a problem.

### Case 1: ergonomics and efficiency

Consider the following (using hypothetical syntax):
```nim
proc floatOrStr(): union(float, string) =
  ...

proc p(): union(int, union(float, string)) =
  return floatOrStr()
```

With the rules as they currently are, this would work as one would expect (in my opinion). If the union type is flattened, then there's a problem, because `union(A, B)` is currently not considered a subtype of `union(A, B, C)`, meaning that the code above becomes illegal.

Of course, `union(A, B)` could be made a subtype of `union(A, B, C)`, and I also think it should. The subtype relationship would - in my opinion - need to be *coercive*, however, given that the necessary run-time conversion is not trivial (the input union needs to be unpacked and then repacked).

In the example showed above, requiring an explicit coercion would hurt ergonomics and also be somewhat counterintuitive.

### Case 2: composition in generics

Consider the following:
```nim
proc f[A, B](a: A, b: B) =
  let x: union(A, B) = select(..., a, b)
  case x
  of A as val: ...
  of B as val: ...
```

My expectation is that a call such as `a: int, b: union(float, string) |- f(a, b)` should be perfectly valid, but when union types are flattened, it wouldn't be, since the second `of` branch would specify a type that's not part of the union.

Again, there are some possible remedies:
1. the notion of *strong* and *weak* union types is introduced; *strong* unions cannot be "destructured", whereas *weak* ones can be. Union types bound to type variables would then always become *strong* ones
2. in `of T as val`, `T` is allowed to be a union type covering a subset of the types in the matched-against union

Number 1 introduces additional complexity to the type system. Number 2 would only work for non-lvalue conversion (because the repacking of unions cannot preserve lvalue-ness), thus having composition issues again.

---

Summing it up, I think that flattening unions (by default) is not a good idea. There could be some type operator, like `flatten`, which constructs a "flat" union type from a nested one, though I think that's something for later.

I think there's also a consistency argument to be made (though maybe not a very strong one) with regards to `tuple`, which doesn't collapse nested tuples, for reasons similar to those I stated above. 

>maybe we shouldn't get a duplication error?

This is also something I though about. I decided to go with reporting an error because I think it'll work better with (future) generic routines.

Consider:
```nim
proc f[A, B](a: A, b: B) =
  let x: union(A, B) = ...
```

When making duplicating types in a union type constructor an error, it's possible to derive an `A != B` constraint for the generic parameters when typing the generic body, making it possible to fail analysis as soon as the concrete types are known, which *could* have some beneficial interactions with overload resolution (should we decide to employ type-based overload resolution, that is).

However, I do have a less strong opinion on an error for duplication than I have on flattening.

Reply:
Thank you for the detailed response. One thing I didn't consider was strong versus weak unions, that's a pretty cool idea; but you're correct I don't see the benefit to cost at present..